### PR TITLE
VAE with linear Normalizing Flows

### PR DESCRIPTION
**Pull request by louis-j-vincent**
Added the VAE with Normalizing Flow model, following the implementation done in https://github.com/mrsalehi/stupid-simple-norm-flow

### CODE REVIEW

**Comment by clementchadebec** [comment]
Hello @louis-j-vincent ,
Thank you very much for opening this PR.
I tried the code but I have an error when running the tests.
```shell
ImportError: cannot import name 'Encoder_VAE_NF_Conv' from 'tests.data.custom_architectures' (/home/clement/Documents/benchmark_VAE/tests/data/custom_architectures.py)
```
It seems this comes from the import of the class `Encoder_VAE_NF_Conv` that does not exist. It should be `Encoder_VAE_Conv` if I am not mistaken. :)

**Review by clementchadebec — COMMENTED** [review]
Hi @louis-j-vincent,
Thank you very much for the new version! I have put some comments on those changes below. Do not hesitate of course if you think otherwise but the main issues that I see so far are:

- [x] The flows are not saved while they are learned. These are needed when reloading the model. Hence, a `save` method should be defined to save the flows and be able able to reload them in `load_from_folder` method likewise the `RHVAE` for instance see [here](https://github.com/clementchadebec/benchmark_VAE/blob/9f7ca87ae6adc45e838646ab8b7dbe9eba509c39/src/pythae/models/rhvae/rhvae_model.py#L258) 
- [x] The tests do not include training, saving and reloading with custom flows. Hence, most of the time the tests that are triggered use a simple vanilla `VAE`   

I have started to integrate some minor comments in this branch [here](https://github.com/clementchadebec/benchmark_VAE/commit/d0d81ff05ab4d839258cdd97b34be7de2fd608d1) maybe this can of any help.

Do not hesitate if you have any questions :)

File: src/pythae/models/vae_nf/vae_nf_config.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from pydantic.dataclasses import dataclass
+
+from typing_extensions import Literal
+
+from ..base.base_config import BaseAEConfig
+
+
+@dataclass
+class VAE_NF_Config(BaseAEConfig):
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_config.py:9]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_config.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    """VAE Normalizing Flow config class.
+
+    Parameters:
+        input_dim (int): The input_data dimension
+        latent_dim (int): The latent space dimension. Default: None.
+        reconstruction_loss (str): The reconstruction loss to use ['bce', 'mse']. Default: 'mse'
+        """
+
+    reconstruction_loss: Literal["bce", "mse"] = "mse"
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_config.py:18]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_model.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        z, eps = self._sample_gauss(mu, std)
+
+        z0 = z.clone().detach()
+
+        log_prob_z0 = torch.sum(
+            -0.5 * torch.log(torch.tensor(2 * math.pi)) - 
+            -0.5 * log_var - 0.5 * ((z0 - mu) / std) ** 2, 
+            axis=-1)
+        
+        log_det = torch.zeros((z0.shape[0],))
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_model.py:117]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_model.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        if "model_config.json" not in file_list:
+            raise FileNotFoundError(
+                f"Missing model config file ('model_config.json') in"
+                f"{dir_path}... Cannot perform model building."
+            )
+
+        path_to_model_config = os.path.join(dir_path, "model_config.json")
+        model_config = VAE_NF_Config.from_json_file(path_to_model_config)
+
+        return model_config
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_model.py:183]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_model.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                    raise NameError(
+                        f"Flow name number {i}: {flow} doesn't correspond to ones of the classes. Available flow classes are [PlanarFlow, RadialFlow]"
+                    )
+
+        self.set_encoder(encoder)
+        self.net = []
+        D = self.decoder.input_dim
+        for flow in flows:
+            layer_class = eval(flow)
+            self.net.append(layer_class(D, torch.tanh))
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_model.py:87]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_model.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            This function requires the folder to contain:
+
+            - | a ``model_config.json`` and a ``model.pt`` if no custom architectures were provided
+
+            **or**
+                
+            - | a ``model_config.json``, a ``model.pt`` and a ``encoder.pkl`` (resp.
+                ``decoder.pkl``) if a custom encoder (resp. decoder) was provided
+
+        """
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_model.py:202]: Done in next commits

File: tests/test_VAE_NF.py [hunk 7]
```diff
@@ -0,0 +1,727 @@
+import os
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply tests/test_VAE_NF.py:1]: Done in next commits

File: tests/test_VAE_NF.py [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+
+import pytest
+import torch
+from torch.optim import SGD, Adadelta, Adagrad, Adam, RMSprop
+
+from pythae.customexception import BadInheritanceError
+from pythae.models.base.base_utils import ModelOutput
+from pythae.models import VAE_NF, VAE_NF_Config
+
+from pythae.trainers import BaseTrainer, BaseTrainingConfig
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply tests/test_VAE_NF.py:12]: Done in next commits

File: src/pythae/models/vae_nf/vae_nf_utils.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        nn.init.normal_(self.z0) 
+        nn.init.normal_(self.log_alpha)
+        nn.init.normal_(self.beta)
+
+
+    def forward(self, z: torch.Tensor):
+        z_sub = z - self.z0
+        alpha = torch.exp(self.log_alpha)
+        r = torch.norm(z_sub)
```
[the remark this answers is not in the document]
Comment (clementchadebec) [reply src/pythae/models/vae_nf/vae_nf_utils.py:56]: Done in next commits

**Comment by louis-j-vincent** [comment]
Resolved import conflicts when merging with updated main branch

**Review by clementchadebec — APPROVED** [review]

**Pull request merged**